test/powershell/Language/Parser/MethodInvocation.Tests.ps1

```diff
@@ -81,3 +81,110 @@ namespace MSFT_716893
+            }
+        )
+
+        $ExpectedParseErrors = @(
+            @{
+                Script          = '$object.Method[incompl'
+                ExpectedErrors  = 'EndSquareBracketExpectedAtEndOfType'
+                ErrorCount      = 1
+            }
+            @{
+                Script          = '[type]::Member[incompl'
+                ExpectedErrors  = 'EndSquareBracketExpectedAtEndOfType'
+                ErrorCount      = 1
+            }
+            @{
+                Script          = '$object.Method[Type1[Type2'
+                ExpectedErrors  = 'UnexpectedToken','EndSquareBracketExpectedAtEndOfType'
+                ErrorCount      = 2
+            }
+        )
+    }
+
+    It 'does not throw a parse error for "<Script>"' -TestCases $EmptyArrayCases {
+        param($Script)
+
+        { [scriptblock]::Create($script) } | Should -Not -Throw
+    }
+
+    It 'reports a parse error for "<Script>"' -TestCases $ExpectedParseErrors {
+        param($Script, $ExpectedError, $ErrorCount)
+
+        $parseErrors = $null
+        [System.Management.Automation.Language.Parser]::ParseInput($Script, [ref]$null, [ref]$parseErrors)
+
+        $parseErrors.Count | Should -Be $ErrorCount
+        $parseErrors.ErrorId | Should -BeIn $ExpectedError
+    }
+
+    It 'can call a generic method "<Script>" with no arguments' -TestCases $EmptyArrayCases {
+        param($Script, $ExpectedType)
+
+        $Result = & [scriptblock]::Create($Script)
+        $Result.GetType() | Should -Be $ExpectedType
+
+        $Result.Length | Should -Be 0
+    }
+
+    It 'can call generic instance methods' {
+        $dictionary = [System.Collections.Concurrent.ConcurrentDictionary[string, int]]::new()
+
+        $addEntryScript = {
+            param($key, $float)
+
+            if ($float -gt 0.5) {
+                return 10
+            }
+            else {
+                return 1
+            }
+        }
+
+        $updateEntryScript = {
+            param($key, $currentValue, $float)
+
+            if ($currentValue / $float -gt 2) {
+                return 5
+            }
+            else {
+                return 0
+            }
+        }
+
+        $FloatValue = 0.4
+        $Key = 'Test'
+
+        # Add entry
+        $dictionary.AddOrUpdate[float]($Key, $addEntryScript, $updateEntryScript, $FloatValue)
+        $dictionary.$Key | Should -Be 1
+
+        # Update entry
+        $dictionary.AddOrUpdate[float]($Key, $addEntryScript, $updateEntryScript, $FloatValue)
+        $dictionary.$Key | Should -Be 5
+    }
+
+    It 'can call generic static methods with arguments' {
+        [System.Linq.Enumerable]::Select[int, int](
+            [int[]](0..10),
+            [func[int, int]] { $args[0] + 2 }
+        ) | Should -Be @(2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)
```

Comment:
It seems we need `Compare-Object -SyncWindow 0` to compare arrays.

Reply:
In what way? `Should` will compare arrays and ensure the order is the same.

Reply:
Doesn't a pipeline enumerate?

Reply:
Yep. `Should` does some extra handling to let it compare arrays from the pipeline. I'm not sure of the implementation details, but `1, 2, 3 | Should -Be 2, 3, 1` will fail, where `1, 2, 3 | Should -Be 1, 2, 3` will pass.

```
PS> 1, 2, 3 | Should -Be 2, 3, 1 
InvalidResult: 
Line |
   1 | 1, 2, 3 | Should -Be 2, 3, 1
     | Expected @(2, 3, 1), but got @(1, 2, 3).

PS> 1, 2, 3 | Should -Be 1, 2, 3
```